src/realm/object-store/sync/app.cpp

```diff
@@ -876,194 +921,226 @@ std::string App::url_for_path(const std::string& path = "") const
     }
+
     Request req;
     req.method = HttpMethod::get;
-    req.url = route;
+    req.url = util::format("%1/location", app_route);
     req.timeout_ms = m_request_timeout_ms;
+    req.redirect_count = redirect_count;
 
-    log_debug("App: request location: %1", route);
-
-    m_config.transport->send_request_to_server(req, [self = shared_from_this(),
-                                                     completion = std::move(completion)](const Response& response) {
-        // If the response contains an error, then pass it up
-        if (response.http_status_code >= 300 || (response.http_status_code < 200 && response.http_status_code != 0)) {
-            return completion(response); // early return
-        }
+    log_debug("App: request location: %1", req.url);
 
-        try {
-            auto json = parse<BsonDocument>(response.body);
-            auto hostname = get<std::string>(json, "hostname");
-            auto ws_hostname = get<std::string>(json, "ws_hostname");
-            auto deployment_model = get<std::string>(json, "deployment_model");
-            auto location = get<std::string>(json, "location");
-            if (self->m_sync_manager->perform_metadata_update([&](SyncMetadataManager& manager) {
-                    manager.set_app_metadata(deployment_model, location, hostname, ws_hostname);
-                })) {
-                // Update the hostname and sync route using the new app metadata info
-                self->update_hostname(self->m_sync_manager->app_metadata());
-            }
-            else {
-                // No metadata in use, update the hostname and sync route directly
-                self->update_hostname(hostname, ws_hostname);
+    m_config.transport->send_request_to_server(
+        std::move(req), [self = shared_from_this(), completion = std::move(completion),
+                         base_url = std::move(base_url)](Request&& request, const Response& response) mutable {
+            // Check to see if a redirect occurred
+            if (AppUtils::is_redirect_status_code(response.http_status_code)) {
+                // Make sure we don't do too many redirects (max_http_redirects (20) is an arbitrary number)
+                if (++request.redirect_count >= s_max_http_redirects) {
+                    completion(AppError{ErrorCodes::ClientTooManyRedirects,
+                                        util::format("number of redirections exceeded %1", s_max_http_redirects),
+                                        {},
+                                        response.http_status_code});
+                    return; // early return
+                }
+                // Handle the redirect response when requesting the location - extract the
+                // new location header field and resend the request.
+                auto redir_location = AppUtils::extract_redir_location(response.headers);
+                if (!redir_location || redir_location->empty()) {
```

Comment:
It can be empty if the headers include this: `{"location": ""}`
And there was already a test for this from before: https://github.com/realm/realm-core/blob/mwb/update-baseurl/test/object-store/sync/app.cpp#L2774-L27

Reply:
Ah, I see. But I don't think an empty location string makes it past the call to `extract_redir_location()` because there  is already a check for the empty string there? Is there any input to `extract_redir_location()` that could produce an empty string? If so, could you add that check to the `SECTION("extract_redir_location")` test?

Reply:
You're right - I removed the extra checks and there is already a test for the empty location (it's the second to last one), but I'm adding a few more tests for other permutations. Thanks